pymilvus/milvus_client/milvus_client.py

```diff
@@ -0,0 +1,801 @@
+
+        return res
+
+    def delete_by_pk(
+        self,
+        pks: list,
+        timeout: int = None,
+    ) -> None:
+        """Delete entries in the collection by their pk.
+
+        Delete all the entries based on the pk. If unsure of pk you can first query the collection
+        to grab the corresponding data. Then you can delete using the pk_field.
+
+        Args:
+            pks (list): _description_
+            timeout (int, optional): Timeout to use, overides the client level assigned at init.
+                Defaults to None.
+        """
+        # TODO: Figure out thread safety
+        # with self.concurrent_lock:
+        #     self.concurrent_counter += 1
+
+        if self.collection is None:
+            logger.error("Collection does not exist: %s", self.collection_name)
+
+        if len(pks) == 0:
+            return
+
+        if self.fields[self.pk_field] == DataType.VARCHAR:
+            ids = ['"' + str(entry) + '"' for entry in pks]
+            expr = f""""{self.pk_field}" in [{','.join(ids)}]"""
+        else:
+            ids = [str(entry) for entry in pks]
+            expr = f"{self.pk_field} in [{','.join(ids)}]"
+
+        self.collection.delete(expr=expr, timout=timeout or self.timeout)
+
+        # TODO: Figure out thread safety
+        # with self.concurrent_lock:
+        #     self.concurrent_counter -= 1
+
+    def delete_collection(
+        self,
+    ) -> None:
+        """Delete the collection."""
+        # TODO: Figure out thread safety
+        with self.concurrent_lock:
+            if self.collection is not None:
+                self.collection.drop()
+                self.collection = None
+
+    def add_partitions(self, input_partitions: List[str]):
+        """Add partitions to the collection.
+
+        Add a list of partition names to the collection. If the collection is loaded
+        it will first be unloaded, then the partitions will be added, and then reloaded.
+
+        Args:
+            input_partitions (List[str]): The list of partition names to be added.
+
+        Raises:
+            MilvusException: Unable to add the partition.
+        """
+        # TODO: Figure out thread safety
+        with self.concurrent_lock:
+            if self.collection is not None and self.is_self_hosted:
+                # Calculate which partitions need to be added
+                input_partitions = set(input_partitions)
+                current_partitions = {
+                    partition.name for partition in self.collection.partitions
+                }
+                new_partitions = input_partitions.difference(current_partitions)
+                # If partitions need to be added, add them
+                if len(new_partitions) != 0:
+                    # Try to unload the collection, if exception already released
+                    reload = False
+                    try:
+                        self.collection.release(timeout=self.timeout)
```

Comment:
Refactor this after https://github.com/milvus-io/milvus/pull/22655 has been released. Then there is no necessary to release the collection when the users tried to add a partition.

Reply:
ah, woops refactored too early.